controllers/machinedeployment_ondelete.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/controllers/mdutil"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// rolloutRolling ....
+func (r *MachineDeploymentReconciler) rolloutOnDelete(ctx context.Context, d *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) error {
+	newMS, oldMSs, err := r.getAllMachineSetsAndSyncRevision(ctx, d, msList, true)
+	if err != nil {
+		return err
+	}
+
+	// newMS can be nil in case there is already a MachineSet associated with this deployment,
+	// but there are only either changes in annotations or MinReadySeconds. Or in other words,
+	// this can be nil if there are changes, but no replacement of existing machines is needed.
+	if newMS == nil {
+		return nil
+	}
+
+	allMSs := append(oldMSs, newMS)
+
+	// Scale up, if we can.
+	if err := r.reconcileNewMachineSetOnDelete(ctx, allMSs, newMS, d); err != nil {
+		return err
+	}
+
+	if err := r.syncDeploymentStatus(allMSs, newMS, d); err != nil {
+		return err
+	}
+
+	// Scale down, if we can.
+	if err := r.reconcileOldMachineSetsOnDelete(ctx, oldMSs, allMSs, d); err != nil {
+		return err
+	}
+
+	if err := r.syncDeploymentStatus(allMSs, newMS, d); err != nil {
+		return err
+	}
+
+	if mdutil.DeploymentComplete(d, &d.Status) {
+		if err := r.cleanupDeployment(ctx, oldMSs, d); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
```

Comment:
This code seems very similar to the other one, should we try to dedupe by passing a function in?

Reply:
Great point: I was thinking we want to keep it separate just so as these grow over time we aren't worried about one code path accidentally causing unforseen changes on another code path but you are totally correct they are similar.

I'm open to whatever you are interested in and can work on a prototype of how this would look

Reply:
The more I look at this the more I think it should stay separate but still open to suggestions.

I did consolidate the reconcileNewMachineSetOnDelete since the logic is identical to that in the rollingUpdate strategy. My worry is someone making a change there that is rolling update specific and breaks OnDelete update in a subtle way. But the benefit is total code is reduced.

I do not necessarily think it's best though to try and define functions that can get passed in for "latest MachineSet" and "all other machine sets" as it seems like that can be limiting.